submariner-operator/issue-79-kubeconfig-handling.md

```diff
@@ -0,0 +1,101 @@
+## Epic Description
+
+The various `subctl` commands support a mixture of `KUBECONFIG` environment variables, `--kubeconfig` flags, `--kubecontext` flags,
+kubeconfig files etc.
+
+We should make this consistent, ideally using `clientcmd`’s tools to handle kubeconfig.
+[Operator PR #23](https://github.com/skitt/submariner-operator/pull/23) has an old attempt at this,
+using [`BindOverrideFlags()`](https://pkg.go.dev/k8s.io/client-go@v0.23.5/tools/clientcmd#BindOverrideFlags) to provide access
+to all the features supported by `clientcmd` (see `kubectl options`).
+
+## Acceptance Criteria
+
+For commands using a single context, _all_ the following should work,
+with kubeconfigs containing one or more clusters and/or contexts (this uses a `kube` prefix when setting up `clientcmd`):
+
+```bash
+# Use the default context in the kubeconfig
+KUBECONFIG=/path/to/kubeconfig subctl foo
+subctl foo --kubeconfig /path/to/kubeconfig
+subctl --kubeconfig /path/to/kubeconfig foo
+
+# Use the specified context from the kubeconfig
+KUBECONFIG=/path/to/kubeconfig subctl foo --kubecontext bar
+subctl foo --kubeconfig /path/to/kubeconfig --kubecontext bar
+subctl --kubeconfig /path/to/kubeconfig --kubecontext bar foo
+```
+
+For commands using multiple contexts, the prefixes should make the purpose of each context clear,
```

Comment:
You asked for this a year or so ago, I’ll try to find the reference.

Reply:
This proposal doesn’t cover turning commands like `benchmark` into matrices. Currently there’s a “from” context and a “to” context, this proposal aims to make that explicit. If we turn `benchmark` into an _n_–_n_ command, we can revisit the context flag naming then.

The goal isn’t to limit commands using flags, but to ensure that the use each command makes of its contexts is explicit. The flag names serve the purposes of the commands, not the other way round; if a command’s behaviour changes, its flag names should also change to suit.

Reply:
Arguably on `benchmark` it actually has two modes - either two context of a single context, for intra-cluster testing, so how would intra-cluster work if I specify just one of the contexts (e.g. just `--fromcontext`)?
Alternatively if I supply multiple configs in `--kubeconfig` and don't specify a context would it fail or not (currently it would if two configs are given for this scenario).

I feel the UX would be simpler if we use just `--kubeconfig` and `--kubecontexts`, or at least if we use same `--kubeconfig` and perhaps several `--[prefix]context` flags.

Reply:
> I feel the UX would be simpler if we use just `--kubeconfig` and `--kubecontexts`, or at least if we use same `--kubeconfig` and perhaps several `--[prefix]context` flags.

My proposal uses the latter approach, see the example given a few lines further down.

As far as `--intra-cluster` handling goes, I don’t know what the best approach is; I think the whole UX here is sub-optimal.